lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -271,6 +271,288 @@ static FailureOr<Literal> parseIntegerLiteral(MLIRContext *context,
+  if (wireInstancePathsFromTop.size() > 1)
+    return wireTarget.ref.getOp()->emitError(
+        "cannot handle multiple paths to DataTaps wire");
+  // Get the path from top to wire
+  auto wirePathFromTop = wireInstancePathsFromTop.back();
+  // A map of the modules in the path from top to wire to its index into the
+  // `wirePathFromTop`.
+  DenseMap<FModuleOp, size_t> wirePathMap;
+  // Initialize the leaf module, to end of path.
+  wirePathMap[wireModule] = wirePathFromTop.size();
+  for (auto &wireInstance : llvm::enumerate(wirePathFromTop))
+    wirePathMap[wireInstance.value()->getParentOfType<FModuleOp>()] =
+        wireInstance.index();
+  auto *wirePathIterator = wirePathFromTop.begin();
+  // Now, reverse iterate over the path of the source, from the source module to
+  // the Top.
+  for (auto srcInstPath : llvm::reverse(srcTarget.instances)) {
+    auto refModule = cast<FModuleOp>(
+        state.instancePathCache.instanceGraph.getReferencedModule(srcInstPath));
+    auto mapIter = wirePathMap.find(refModule);
+    // If `refModule` exists on the wire path, then this is the Lowest Common
+    // Ancestor between the source and wire module.
+    if (mapIter != wirePathMap.end()) {
+      lcaModule = refModule;
+      wirePathIterator += mapIter->getSecond();
+      break;
+    }
+    pathFromSrcToWire.push_back(srcInstPath);
+  }
+  pathFromSrcToWire.insert(pathFromSrcToWire.end(), wirePathIterator,
+                           wirePathFromTop.end());
+
+  return success();
+}
+
+LogicalResult static applyNoBlackBoxStyleDataTaps(const AnnoPathValue &target,
+                                                  DictionaryAttr anno,
+                                                  ApplyState &state) {
+  auto *context = state.circuit.getContext();
+  auto loc = state.circuit.getLoc();
+
+  // Process all the taps.
+  auto keyAttr = tryGetAs<ArrayAttr>(anno, anno, "keys", loc, dataTapsClass);
+  if (!keyAttr)
+    return failure();
+  for (size_t i = 0, e = keyAttr.size(); i != e; ++i) {
+    auto b = keyAttr[i];
+    auto path = ("keys[" + Twine(i) + "]").str();
+    auto bDict = b.cast<DictionaryAttr>();
+    auto classAttr =
+        tryGetAs<StringAttr>(bDict, anno, "class", loc, dataTapsClass, path);
+    if (!classAttr)
+      return failure();
+    // Can only handle ReferenceDataTapKey and DataTapModuleSignalKey
+    if (classAttr.getValue() != referenceKeyClass &&
+        classAttr.getValue() != internalKeyClass)
+      return mlir::emitError(loc, "Annotation '" + Twine(dataTapsClass) +
+                                      "' with path '" + path + ".class" +
+                                      +"' contained an unknown/unimplemented "
+                                       "DataTapKey class '" +
+                                      classAttr.getValue() + "'.")
+                 .attachNote()
+             << "The full Annotation is reprodcued here: " << anno << "\n";
```

Comment:
This stuff looks like a lot of duplicated code from the non-blackbox applicator. This is fine as long as we do the migration relatively quickly to the new non-blackbox format. Otherwise, it would be good to factor out the common pieces into static functions (or some other mechanism that avoids duplication).

Reply:
Yes, I want to completely remove the DataTaps and MemTaps files. Just move the implementation to `GrandCentral.cpp` , they all share a lot of this logic.